### PR TITLE
HenryJia: auto-move data decorator

### LINKED ISSUE #1412: Auto move input to proper device for inference
Does PyTorch Lightning provide abstractions for inference? In particular, does it provide ways of automatically handling the transfer to/from GPU when I call `model(x)`, or do I need to roll my own code for that?

## Example Use Case

I have a use case where I train a model on slices of a sliding window of an audio spectrogram (i.e., let's say 1 second chunks). When training is finished, I'd like to see the performance of the model on an entire file. Pseudocode:

```python
# generate training data
X, Y = [], []
for audio_file in audio_files:
    for x, y in sliding_window(audio_file):
        X.append(x); Y.append(y)
X, Y = shuffle(X, Y)  # shuffle the slices of all files

# Train model on slices
model = ExampleModel(X, Y)
trainer = Trainer(gpus=1)
trainer.fit(model)

# Plot the performance on a whole test file:
test_Y = []
for x, _ in sliding_window(test_file)
    test_Y.append(model(x))
plt.plot(test_Y)
```

Notice that during training, the notion of a file is entirely gone, but when I plot my test file, I reintroduce that. Of course, in my real code, my training data `X, Y` is split into training, validation and test, as usual. The plotting step is an additional verification; sort of like putting the pieces together.

## Problem
When the model runs on the GPU, The last part of the code becomes:
```python
# Plot the performance on a whole test file:
model.eval()
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
test_Y = []
for x, _ in sliding_window(test_file)
    y = model(x.to(device)).cpu()
    test_Y.append(y)
plt.plot(test_Y)
```

This isn't the end of the world, but it's not as nice as the other code that PyTorch Lightning helped me refactor. I also can't call `x.type_as(...)` since in that loop, I have no reference type that lives on the CPU/GPU that I could refer to (or maybe I can, but I haven't figured it out).

A workaround to this is to save the model and load it again, on a CPU.

```python
# Train model on slices
# ...
trainer.fit(model)
trainer.save_checkpoint("model.ckpt")
model = ExampleModel.load_from_checkpoint("model.ckpt")

# Plot the performance on a whole test file:
model.eval()
test_Y = []
for x, _ in sliding_window(test_file)
    test_Y.append(model(x))
plt.plot(test_Y)
```

While this removes the noise of the `.to(device)` and `.cpu()` calls, it adds the overhead of having to save the model every time. I also still have to manually call `model.eval()`. The use case of running my model on an entire audio file is not for metrics but for visual inspection; as such I always only sample a few audio files. Running the model on a CPU instead of a GPU for inference thus isn't a problem.

## Question
Is there a more elegant way to achieve the above?

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?
Continuation of the work of @HenryJia from PR #1526, fixes #1412
This is currently based on PR #1756.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@awaelchli seems this is unblocked now, mind resolve conflicts? :rabbit: 
@HenryJia mind check the update?

**Comment by awaelchli** [comment]
Thanks for checking @HenryJia . I rebased and tests run except for one. 
I added a doctest but it requires the GPU. It fails if the machine only has CPU. 
I'm looking into that, I would really like to have that doctest. But otherwise PR is ready.

**Comment by Borda** [comment]
> I'm looking into that, I would really like to have that doctest. But otherwise PR is ready.

I would like to have the doctest for it too, but maybe it would be easier to have the particular code just as an example and get this merge and later try to find a way to have it as doctest...

File: tests/core/test_decorators.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+def test_auto_move_data(src_device, dest_device):
+    """ Test that the decorator moves the data to the device the model is on. """
+
+    class CurrentModel(EvalModelTemplate):
+        pass
+
+    # apply the decorator
+    CurrentModel.forward = auto_move_data(CurrentModel.forward)
+
+    model = CurrentModel()
```
Comment (Borda) [inline tests/core/test_decorators.py:24]:
```suggestion
    model = EvalModelTemplate()
    # apply the decorator
    model.forward = auto_move_data(model.forward)
```
why pure `EvalModelTemplate` is not enough and you need a child?

Comment (awaelchli) [reply tests/core/test_decorators.py:24]: unfortunately not working. the decorator needs to be applied to the class definition. 
So we can do `EvalModelTemplate.forward = auto_move_data(EvalModelTemplate.forward)` but then it is applied for all following tests too.

Comment (Borda) [reply tests/core/test_decorators.py:24]: yeah, that looks easier :rabbit:

Comment (awaelchli) [reply tests/core/test_decorators.py]: this has side effects on other tests (and are failing). the decorator should only be applied at test time, when we e.g. only have one gpu. 
I think it's cleaner if we applied the decorator only locally in this test.

**Comment by Borda** [comment]
@HenryJia @awaelchli could we get this done? :raccoon:

**Comment by awaelchli** [comment]
@Borda If the decorator is enough, I think it is done module review. However this is probably not going to be merged very soon since there are these huge PRs on the top of the list.

**Comment by awaelchli** [comment]
@SkafteNicki thanks, good suggestion. I added a short description to the deployment section.

**Comment by awaelchli** [comment]
Hey @HenryJia we made it!! good work ✋

**Comment by HenryJia** [comment]
Awesome cheers <3 thanks for continuing it when I was too busy

**Pull request merged**